### Import Libraries

In [6]:
import pandas as pd
import matplotlib as plt
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import mean_squared_error, r2_score, balanced_accuracy_score
from sklearn.metrics import confusion_matrix, classification_report, ConfusionMatrixDisplay
import os

### Load Clean DataSet

In [7]:
df=pd.read_csv('clean-healthcare-dataset-stroke-data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4908 entries, 0 to 4907
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   age                             4908 non-null   float64
 1   hypertension                    4908 non-null   int64  
 2   heart_disease                   4908 non-null   int64  
 3   avg_glucose_level               4908 non-null   float64
 4   bmi                             4908 non-null   float64
 5   Male                            4908 non-null   int64  
 6   ever_married                    4908 non-null   int64  
 7   Urban                           4908 non-null   int64  
 8   work_type_Govt_job              4908 non-null   int64  
 9   work_type_Never_worked          4908 non-null   int64  
 10  work_type_Private               4908 non-null   int64  
 11  work_type_Self-employed         4908 non-null   int64  
 12  work_type_children              49

## About the DataSet

### **EXTREMELY UNBALANCED**

### The "cleaned and encoded" dataset contains 4908 records
### Only 209 (4.25%) of these are positive for stroke and the rest (95.74%) are negative

### Will try to create models with different less unbalanced samples to run prediction models

### **Fortunately/Unfortunately there might not be enough positive cases to effectively predict**

### Training Models

In [8]:
models={}
models['Logistic Regression']=LogisticRegression()
models['Random Forest Clasifier']=RandomForestClassifier()
models['SVC']=SVC()

### Re-Assign and int value to the Scaled result

### DataFrame Dictionary with Proportional Positive and negative Samples

In [9]:

def create_sample_dfs_dictionary(df):
    Positives=df.loc[df['stroke']==df['stroke'].max()]
    Negative_sample_1=df.loc[df['stroke']==df['stroke'].min()] \
        .sample(len(Positives),random_state=1, ignore_index=True)
    Negative_sample_2=df.loc[df['stroke']==df['stroke'].min()] \
        .sample(len(Positives),random_state=3, ignore_index=True)
    Negative_sample_3=df.loc[df['stroke']==df['stroke'].min()] \
        .sample(len(Positives),random_state=5, ignore_index=True)
    DataFrames={}
    DataFrames['One to One']= pd.concat([Positives,Negative_sample_1], axis=0)
    DataFrames['One to Two']= pd.concat([Positives,Negative_sample_1, \
                                        Negative_sample_2], axis=0)
    DataFrames['One to Three']= pd.concat([Positives,Negative_sample_1, \
                                        Negative_sample_2,Negative_sample_3], axis=0)
    return DataFrames

### Print Explained Confusion Matrix

In [10]:
def print_confusion_matrix(y_test,prediction,Labels):
    cm=confusion_matrix(y_test,prediction,labels=Labels)
    print(f'\033[1mConfusion Matrix\033[0m')
    print(f'Predicted\\Actual\tPositive\tNegative')
    print(f'Positive({cm[0][0]+cm[0][1]})\t\t{cm[0][0]}\t\t{cm[0][1]}')
    print(f'Negative({cm[1][0]+cm[1][1]})\t\t{cm[1][0]}\t\t{cm[1][1]}')

### RandomForestClassifier best depth calculator

In [11]:
def best_depth(X_train,X_test,y_train,y_test):
    # print('Calculating best depth for RandomForestClassifier')
    train_score=0
    depth=0
    best=0
    hight_score=0
    while train_score<1:
        depth+=1
        clf = RandomForestClassifier(max_depth = depth)
        clf.fit(X_train, y_train)
        train_score = clf.score(X_train,y_train)
        test_score = clf.score(X_test,y_test)
        score=test_score*(1-(train_score-test_score))
        if hight_score<score:
            hight_score=score
            best=depth
    return best      


In [12]:
# DataFrames=create_sample_dfs_dictionary(df)
for name, Learning_model in models.items():
    print(name)

Logistic Regression
Random Forest Clasifier
SVC


### **train_test_split and ALL the tests**

In [35]:
def Test_Results(df):
    X_available=df.copy().drop(columns='stroke')
    y_available=df['stroke']
    tab='   '
    DataFrames=create_sample_dfs_dictionary(df)
    for name, Learning_model in models.items():    
        print(f'\033[94m\033[1m{name}\033[0m')
        for key,df in DataFrames.items():
            X=df.copy().drop(columns='stroke')
            y=df['stroke']
            X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=1)
            if name=='Random Forest Clasifier':
                Depth=best_depth(X_train,X_test,y_train,y_test)
                Learning_model=RandomForestClassifier(max_depth=Depth)
            model=Learning_model
            model.fit(X_train,y_train)
            predicted = model.predict(X_test)
            available_predict=model.predict(X_available)

            # Score the predictions with mse and r2
            test_balanced_accuracy=balanced_accuracy_score(y_test, predicted)
            available_balanced_accuracy=balanced_accuracy_score(y_available, available_predict)
            tcr=classification_report(y_test, predicted, labels=[1,0],output_dict=True)
            acr=classification_report(y_available, available_predict,
                                      labels=[1,0],output_dict=True)
            #region Print Results
            if name=='Random Forest Clasifier':
                print(f'{tab}\033[96m\033[1m{key} ({len(df)} items, depth of {Depth})\033[0m : ', end='')
            else:
                print(f'{tab}\033[96m\033[1m{key} ({len(df)} items)\033[0m : ', end='')
            print(f'Model Score: {model.score(X_train,y_train):.4f}')
            print(f'\t\t\t\t\t\033[1mTest Data\tAvailable Data\t\tDifference\033[0m')
            print(f"{tab}{tab}Balanced Accuracy:\t\t{test_balanced_accuracy:.4f}"+
                f"\t\t\t{available_balanced_accuracy:.4f}\t\t"+
                f"{abs(test_balanced_accuracy-available_balanced_accuracy):.4f}")
            print(f"{tab}{tab}\033[1mNegative(0) results metrics\033[0m")
            print(f"{tab}{tab}{tab}Precision\t\t\t{tcr['0']['precision']:.4f}\t\t\t"+
                  f"{acr['0']['precision']:.4f}\t\t"+
                  f"{(tcr['0']['precision']-acr['0']['precision']):.4f}")
            print(f"{tab}{tab}{tab}Recall\t\t\t\t{tcr['0']['recall']:.4f}\t\t\t"+
                  f"{acr['0']['recall']:.4f}\t\t"+
                  f"{(tcr['0']['recall']-acr['0']['recall']):.4f}")
            print(f"{tab}{tab}\033[1mPositive(1) results metrics\033[0m")
            print(f"{tab}{tab}{tab}Precision\t\t\t{tcr['1']['precision']:.4f}\t\t\t"+
                  f"{acr['1']['precision']:.4f}\t\t"+
                  f"{(tcr['1']['precision']-acr['1']['precision']):.4f}")
            print(f"{tab}{tab}{tab}Recall\t\t\t\t{tcr['1']['recall']:.4f}\t\t\t"+
                  f"{acr['1']['recall']:.4f}\t\t"+
                  f"{(tcr['1']['recall']-acr['1']['recall']):.4f}")
            
            # cm=confusion_matrix(y_test, predicted,[1,0])
            # disp=Confu

            print()
            #endregion
            
        print('***'*10)

### StandardScaler Scaled

In [14]:
ss=StandardScaler()
ss_scaled=ss.fit_transform(df)
ss_df=pd.DataFrame(ss_scaled, columns=df.columns)
ss_df

age  hypertension  heart_disease  avg_glucose_level       bmi  \
0     1.069938     -0.318102       4.381499           2.777797  0.981145   
1     1.646336     -0.318102       4.381499           0.014016  0.459086   
2     0.271847     -0.318102      -0.228232           1.484266  0.701016   
3     1.601998      3.143642      -0.228232           1.549325 -0.623231   
4     1.690675     -0.318102      -0.228232           1.821493  0.013426   
...        ...           ...            ...                ...       ...   
4903 -1.324334     -0.318102      -0.228232          -0.049918 -1.310821   
4904  1.690675     -0.318102      -0.228232           0.448045  1.414072   
4905 -0.348890     -0.318102      -0.228232          -0.502181  0.217156   
4906  0.360524     -0.318102      -0.228232           1.373057 -0.419501   
4907  0.050155     -0.318102      -0.228232          -0.450629 -0.343102   

          Male  ever_married     Urban  work_type_Govt_job  \
0     1.200240      0.729270  0.985436           -0.383751   
1     1.200240      0.729270 -1.014779           -0.383751   
2    -0.833166      0.729270  0.985436           -0.383751   
3    -0.833166      0.729270 -1.014779           -0.383751   
4     1.200240      0.729270  0.985436           -0.383751   
...        ...           ...       ...                 ...   
4903 -0.833166     -1.371234 -1.014779           -0.383751   
4904 -0.833166      0.729270  0.985436           -0.383751   
4905 -0.833166      0.729270 -1.014779           -0.383751   
4906  1.200240      0.729270 -1.014779           -0.383751   
4907 -0.833166      0.729270  0.985436            2.605854   

      work_type_Never_worked  work_type_Private  work_type_Self-employed  \
0                  -0.067102           0.864071                -0.433030   
1                  -0.067102           0.864071                -0.433030   
2                  -0.067102           0.864071                -0.433030   
3                  -0.067102          -1.157312                 2.309308   
4                  -0.067102           0.864071                -0.433030   
...                      ...                ...                      ...   
4903               -0.067102          -1.157312                -0.433030   
4904               -0.067102          -1.157312                 2.309308   
4905               -0.067102          -1.157312                 2.309308   
4906               -0.067102           0.864071                -0.433030   
4907               -0.067102          -1.157312                -0.433030   

      work_type_children  smoking_status_Unknown  \
0              -0.397953               -0.658022   
1              -0.397953               -0.658022   
2              -0.397953               -0.658022   
3              -0.397953               -0.658022   
4              -0.397953               -0.658022   
...                  ...                     ...   
4903            2.512858                1.519706   
4904           -0.397953               -0.658022   
4905           -0.397953               -0.658022   
4906           -0.397953               -0.658022   
4907           -0.397953                1.519706   

      smoking_status_formerly smoked  smoking_status_never smoked  \
0                           2.206992                    -0.778473   
1                          -0.453105                     1.284565   
2                          -0.453105                    -0.778473   
3                          -0.453105                     1.284565   
4                           2.206992                    -0.778473   
...                              ...                          ...   
4903                       -0.453105                    -0.778473   
4904                       -0.453105                     1.284565   
4905                       -0.453105                     1.284565   
4906                        2.206992                    -0.778473   
4907                       -0.453105                    

### Re-assign 1 and 0 to the value results

In [15]:
ss_df['stroke']=ss_df['stroke'].apply(lambda x: 1 if x>0 else 0)
ss_df['stroke'].value_counts()

stroke
0    4699
1     209
Name: count, dtype: int64

In [36]:
Test_Results(ss_df)

Logistic Regression
   One to One (418 items) : Model Score: 0.7572
					Test Data	Available Data		Difference
      Balanced Accuracy:		0.8524			0.7576		0.0948
      Negative(0) results metrics
         Precision			0.9057			0.9883		-0.0827
         Recall				0.8136			0.7019		0.1117
      Positive(1) results metrics
         Precision			0.7885			0.1082		0.6803
         Recall				0.8913			0.8134		0.0779

   One to Two (627 items) : Model Score: 0.7894
					Test Data	Available Data		Difference
      Balanced Accuracy:		0.7272			0.7406		0.0134
      Negative(0) results metrics
         Precision			0.7946			0.9814		-0.1867
         Recall				0.8725			0.8400		0.0326
      Positive(1) results metrics
         Precision			0.7111			0.1512		0.5599
         Recall				0.5818			0.6411		-0.0593

   One to Three (836 items) : Model Score: 0.7974
					Test Data	Available Data		Difference
      Balanced Accuracy:		0.7239			0.7069		0.0170
      Negative(0) results metrics
         Precision			0.8606			0

### MinMaxScaler

In [17]:
mms=MinMaxScaler()
mms_scaled=mms.fit_transform(df)
mms_df=pd.DataFrame(mms_scaled, columns=df.columns)
mms_df

age  hypertension  heart_disease  avg_glucose_level       bmi  \
0     0.816895           0.0            1.0           0.801265  0.301260   
1     0.975586           0.0            1.0           0.234512  0.254296   
2     0.597168           0.0            0.0           0.536008  0.276060   
3     0.963379           1.0            0.0           0.549349  0.156930   
4     0.987793           0.0            0.0           0.605161  0.214204   
...        ...           ...            ...                ...       ...   
4903  0.157715           0.0            0.0           0.221402  0.095074   
4904  0.987793           0.0            0.0           0.323516  0.340206   
4905  0.426270           0.0            0.0           0.128658  0.232532   
4906  0.621582           0.0            0.0           0.513203  0.175258   
4907  0.536133           0.0            0.0           0.139230  0.182131   

      Male  ever_married  Urban  work_type_Govt_job  work_type_Never_worked  \
0      1.0           1.0    1.0                 0.0                     0.0   
1      1.0           1.0    0.0                 0.0                     0.0   
2      0.0           1.0    1.0                 0.0                     0.0   
3      0.0           1.0    0.0                 0.0                     0.0   
4      1.0           1.0    1.0                 0.0                     0.0   
...    ...           ...    ...                 ...                     ...   
4903   0.0           0.0    0.0                 0.0                     0.0   
4904   0.0           1.0    1.0                 0.0                     0.0   
4905   0.0           1.0    0.0                 0.0                     0.0   
4906   1.0           1.0    0.0                 0.0                     0.0   
4907   0.0           1.0    1.0                 1.0                     0.0   

      work_type_Private  work_type_Self-employed  work_type_children  \
0                   1.0                      0.0                 0.0   
1                   1.0                      0.0                 0.0   
2                   1.0                      0.0                 0.0   
3                   0.0                      1.0                 0.0   
4                   1.0                      0.0                 0.0   
...                 ...                      ...                 ...   
4903                0.0                      0.0                 1.0   
4904                0.0                      1.0                 0.0   
4905                0.0                      1.0                 0.0   
4906                1.0                      0.0                 0.0   
4907                0.0                      0.0                 0.0   

      smoking_status_Unknown  smoking_status_formerly smoked  \
0                        0.0                             1.0   
1                        0.0                             0.0   
2                        0.0                             0.0   
3                        0.0                             0.0   
4                        0.0                             1.0   
...                      ...                             ...   
4903                     1.0                             0.0   
4904                     0.0                             0.0   
4905                     0.0                             0.0   
4906                     0.0                             1.0   
4907                     1.0                             0.0   

      smoking_status_never smoked  smoking_status_smokes  stroke  
0                             0.0                    0.0     1.0  
1                             1.0                    0.0     1.0  
2                             0.0                    1.0     1.0  
3                             1.0                    0.0     1.0  
4                             0.0                    0.0     1.0  
...                           ...                    ...     ...  
4903                          0.0             

In [18]:
Test_Results(mms_df)

Logistic Regression
   One to One (418 items) : Model Score: 0.7572
					Test Data	Available Data		Difference
      Balanced Accuracy:		0.8185		0.7583			0.0602
      Precision on 0 		0.8979591836734694		0.9890009165902841

   One to Two (627 items) : Model Score: 0.7957
					Test Data	Available Data		Difference
      Balanced Accuracy:		0.7510		0.7489			0.0021
      Precision on 0 		0.8070175438596491		0.9817073170731707

   One to Three (836 items) : Model Score: 0.7879
					Test Data	Available Data		Difference
      Balanced Accuracy:		0.6771		0.6499			0.0271
      Precision on 0 		0.8306010928961749		0.9707252162341983

******************************
Random Forest Clasifier
   One to One (418 items, depth of 4) : Model Score: 0.8051
					Test Data	Available Data		Difference
      Balanced Accuracy:		0.8148		0.7739			0.0409
      Precision on 0 		0.9318181818181818		0.9945762711864407

   One to Two (627 items, depth of 3) : Model Score: 0.8000
					Test Data	Available Data		Differen